SwipeStripe Review PR (do not merge)

tractorcow opened the pull request:
Review PR for leaving comments on development of this module.

hchokshi committed:
Add alteration message when required page is created

hchokshi committed:
Prevent un-publishing of required pages

hchokshi committed:
Use DBPrice injector spec constant over class ref for create_field

hchokshi committed:
Use translation entities over hardcoded messages

hchokshi committed:
Move active cart trait into Pages namespace

hchokshi committed:
Add view cart page with quantity updating form

hchokshi committed:
Fix quantity setter on OrderItem

hchokshi committed:
Add more lock checks on order item

hchokshi committed:
Add stock availability check on OrderItemQuantityField

hchokshi committed:
Prevent Order generating OrderItems if immutable

hchokshi committed:
Extract active cart session key to SessionData constants

hchokshi committed:
Use redirect to hide order guest token from URL

hchokshi committed:
Show add-ons, unit price info in cart quantity field

hchokshi committed:
Show cart subtotal, total, order add-ons

hchokshi committed:
Add unique index to customer email column

hchokshi committed:
Add orders method to member extension

hchokshi committed:
Add message to cart page if checkout lock active

hchokshi committed:
Fix setQuantity on new OrderItem

hchokshi committed:
Add summary fields to add-ons, Order and OrderItem

hchokshi committed:
Add permissions for viewing/editing orders

hchokshi committed:
Remove available count from PurchasableInterface

Let implementors handle stock management etc. if wanted.

hchokshi committed:
Add lock/unlock logic to Order/OrderItem

hchokshi committed:
Add checkout page and link on non-empty cart page

hchokshi committed:
Explicitly set cart form action

hchokshi committed:
Use ActiveCart as property for caching benefits

hchokshi committed:
Make order item purchasable nullable

hchokshi committed:
Allow double-calling of order lock/unlock

hchokshi committed:
Add order confirmation page

hchokshi committed:
Fix Order relation call in PaymentExtension

hchokshi committed:
Extract cart total summary template into include

hchokshi committed:
Delete empty OrderItems at checkout time

Allows users to see effects of removing an item without having to go off
cart screen to add it back.

hchokshi committed:
Move AddOn trait into Order namespace

hchokshi committed:
Move cart quantity readonly logic into field

hchokshi committed:
Simplify cart immutable validation logic

hchokshi committed:
Add remove button to order quantity field

hchokshi committed:
Implement preliminary order payment capture/cancel behaviour

hchokshi committed:
Add link method to order

hchokshi committed:
Add hash method to order for verifying content changes

hchokshi committed:
Add unpaid total method to Order

hchokshi committed:
Add order methods to add/remove items

hchokshi committed:
Add checkout form and template

hchokshi committed:
Ensure cart always unlocked for checkout

hchokshi committed:
Show more detailed errors for payment failure

hchokshi committed:
Remove trait config dependency

Trait config transformer cannot be injected into SilverStripe’s 
TestKernel without editing the source.

hchokshi committed:
Comment test case failing due to configuration mismatch

hchokshi committed:
Add coverage logging to travis config

hchokshi committed:
Update silverstripe/framework constraint to match CMS constraint

hchokshi committed:
Restore original author to composer.json authors key

hchokshi committed:
Add name headers to YAML config

hchokshi committed:
Remove editor specific ignores from .gitignore

hchokshi committed:
Add recipe-cms as dev dependency instead of TestOnly Page/Controller

hchokshi committed:
Merge pull request #2 from hchokshi/5.0

Merge initial stage of SilverStripe 4 rewrite

hchokshi committed:
Update templates for CWP Watea layout

hchokshi committed:
Merge pull request #3 from hchokshi/feature/cwp-watea-templates

Update templates for CWP Watea layout

hchokshi committed:
Add CI build and coverage badges to readme

hchokshi committed:
Merge pull request #4 from hchokshi/5.0

Add CI build and coverage badges to readme

hchokshi committed:
Move PurchasableInterface into Order namespace

hchokshi committed:
Move PurchasableInterface into Order namespace

hchokshi committed:
Rename test suite to “SwipeStripe Core”

hchokshi committed:
Add test purchasable data object

hchokshi committed:
Move coverage from PHPUnit config to travis config

hchokshi committed:
Use php-dbg to generate coverage data

PHPDBG is faster generating coverage than XDebug.
https://github.com/sebastianbergmann/phpunit/issues/1662#issuecomment-405745628

hchokshi committed:
Use dist composer install wherever possible

Dist installs are MUCH faster, but source installs are required for 
SilverStripe test bootstrap code to be loaded.

hchokshi committed:
Add unit tests for cart functionality

hchokshi committed:
Add Page/PageController to autoload-dev

hchokshi committed:
Fix OrderItem quantity write logic

hchokshi committed:
Add tests for order empty and hash

hchokshi committed:
Fix logic for order empty check

hchokshi committed:
Add Omnipay dummy gateway as dev dependency

hchokshi committed:
Add test for order unpaid total calculation

hchokshi committed:
Add empty checks before applying add-ons

hchokshi committed:
Fix add-on test, add test for OrderItem add-ons

hchokshi committed:
Merge pull request #5 from hchokshi/feature/purchasable-in-order-namespace

Move PurchasableInterface into Order namespace

hchokshi committed:
Add tests for member/customer-order relationship

hchokshi committed:
Use fixtures to create pages for tests

hchokshi committed:
Replace test product with fixture

hchokshi committed:
Add test for eCommerce default page records

hchokshi committed:
Extract product fixture as class property

hchokshi committed:
Add test for detaching order item add-ons

hchokshi committed:
Ensure sub-classes not included in get_one calls

hchokshi committed:
Add test for Order link generation

hchokshi committed:
Add fixtures for guest and account customer

hchokshi committed:
Add tests for supported currencies

hchokshi committed:
Merge pull request #7 from hchokshi/feature/improve-test-coverage

Improve test coverage

hchokshi committed:
Rename Fixtures::BASE_PATH constant to FIXTURE_BASE_PATH

hchokshi committed:
Add test for AmountField

hchokshi committed:
Add more tests for PriceField

hchokshi committed:
Add tests for Payable extension

hchokshi committed:
Add test for order version locking

hchokshi committed:
Fix CMS error loading ModelAdmin with add-ons

hchokshi committed:
Merge pull request #8 from hchokshi/feature/improve-test-coverage

Improve test coverage

hchokshi committed:
Update cart form for Watea theme

hchokshi committed:
Update checkout form for Watea theme

hchokshi committed:
Redesign order / confirmation pages for Watea theme

hchokshi committed:
Simplify PriceField template

hchokshi committed:
Remove Customer object

hchokshi committed:
Add country field dependency

hchokshi committed:
Add DBAddress composite field

hchokshi committed:
Add customer details to Order object

hchokshi committed:
Save customer details into order at checkout

hchokshi committed:
Send confirmation email on order completion

hchokshi committed:
Display country name instead of code in Address.Nice

hchokshi committed:
Hide internal state from Order CMS

hchokshi committed:
Merge pull request #9 from hchokshi/feature/templates-watea

Update existing templates for CWP Watea theme

hchokshi committed:
Add customer details to order page

hchokshi committed:
Add customer name and email to order summary fields

hchokshi committed:
Add order model admin for non-cart orders

hchokshi committed:
Add ReadOnlyGridField to show useful immutable grid fields

hchokshi committed:
Add inline data about purchased item to order item CMS

hchokshi committed:
Add view/edit permission checks based on Order

hchokshi committed:
Update payment CMS fields to show transaction status/reference

hchokshi committed:
Clean up order/item add-on CMS view

hchokshi committed:
Make totals more visible in order CMS

hchokshi committed:
Auto-populate order fields with current account

hchokshi committed:
Extract read-only grid field creator into helper

hchokshi committed:
Fix DBField copy bug

hchokshi committed:
Re-order Order CMS tabs

hchokshi committed:
Add created/edited times to Order CMS

hchokshi committed:
Fix searchable fields error in OrderItem

hchokshi committed:
Add price to OrderItem CMS

hchokshi committed:
Add view orders as required permission code for OrderAdmin

hchokshi committed:
Fix test failures

hchokshi committed:
Merge pull request #10 from hchokshi/feature/checkout-billing-details

Feature - billing details at checkout

hchokshi committed:
Merge pull request #11 from hchokshi/feature/admin-view-orders

Feature - admin view orders

hchokshi committed:
Remove Silverstripe Omnipay fork from dependencies

Upstream merged PR with Omnipay 3 updates.

hchokshi committed:
Merge pull request #12 from hchokshi/feature/remove-omnipay-fork

Switch Silverstripe Omnipay dependency from fork to upstream

hchokshi committed:
Add absolute link method to Order

Fixes email view order link issue.

hchokshi committed:
Add WKHTML dependency (SS4 upgraded fork)

hchokshi committed:
Add confirmation time to order

hchokshi committed:
Move order template includes into Order directory

hchokshi committed:
Extract payment summary template into include

hchokshi committed:
Extract order or 404 logic into method

hchokshi committed:
Add ability to download receipt as PDF

hchokshi committed:
Add environment type to order record

hchokshi committed:
Simplify order permission translation strings

hchokshi committed:
Add permissions for products

hchokshi committed:
Merge pull request #14 from hchokshi/feature/product-permissions

Add product permissions

hchokshi committed:
Wrap all CMS field generation in beforeExtends call

hchokshi committed:
Merge pull request #15 from hchokshi/feature/before-extend-cms-fields

Wrap all CMS field generation in beforeExtends call

hchokshi committed:
Merge pull request #13 from hchokshi/feature/download-receipt

Feature - download receipt

hchokshi committed:
Move OrderAdmin into Order namespace

hchokshi committed:
Move supported currencies into Price namespace

hchokshi committed:
Move ReadOnlyGridField into Forms namespace

hchokshi committed:
Move OrderItemQuantityField into OrderItem namespace

hchokshi committed:
Move AddOnPriority/PaymentStatus into Order namespace

hchokshi committed:
Move SessionDate/ShopPermissions into root SwipeStripe namespace

hchokshi committed:
Remove BaseForm class

hchokshi committed:
Move cart/checkout forms into Order namespace

hchokshi committed:
Move OrderConfirmationEmail into Order namespace

hchokshi committed:
Merge pull request #16 from hchokshi/feature/reorganise-namespaces

Reorganise namespaces

hchokshi committed:
Move HasActiveCart into SwipeStripe root namespace

hchokshi committed:
Move checkout form and pages into Order\Checkout namespace

hchokshi committed:
Move cart form and page into Order\Cart namespace

hchokshi committed:
Move view order and confirmation into Order namespace

hchokshi committed:
Move RequiredSinglePage trait out of Pages namespace

hchokshi committed:
Update page test fixtures

hchokshi committed:
Merge pull request #17 from hchokshi/feature/reorganise-pages

Reorganise pages

hchokshi committed:
Extract cart form validation into validator class

hchokshi committed:
Extract checkout form validation into validator class

hchokshi committed:
Handle guest cart logging in before checkout

hchokshi committed:
Simplify payment identifier verification logic

hchokshi committed:
Use Controller::join_links over string concatenation

hchokshi committed:
Merge pull request #18 from hchokshi/feature/extract-form-validators

Extract form validators into validator classes

hchokshi committed:
Simplify checkout form validator

hchokshi committed:
Show guest users option to create an account

hchokshi committed:
Add password field for creating account during checkout

hchokshi committed:
Use dependency injection for payment service factory

hchokshi committed:
Default to creating account for guest checkout

hchokshi committed:
Replace country dropdown field module with simpler implementation

hchokshi committed:
Merge pull request #19 from hchokshi/feature/replace-country-select-field

Replace country dropdown field module with simpler implementation

hchokshi committed:
Replace checkout form accounts logic with extension points

hchokshi committed:
Ensure payment error gets stripped on submission validation failure

hchokshi committed:
Extract SwipeStripe email super-class

hchokshi committed:
Merge pull request #20 from hchokshi/feature/create-account

Add hooks for checkout account creation (swipestripe/accounts module)

hchokshi committed:
Use $Layout in email templates to allow shared base structure

hchokshi committed:
Add empty check method to DBAddress

hchokshi committed:
Move account defaults into cart functionality to accounts module

hchokshi committed:
Merge pull request #21 from hchokshi/feature/default-address-to-module

Move default billing address to accounts module

hchokshi committed:
Allow DBAddress to be populated from form data array

hchokshi committed:
Remove null type-hint from has_one relations

No relation returns a mock object with ID 0, not null.

hchokshi committed:
Move checkout form member logic from core to accounts

hchokshi committed:
Update PHPDoc hints

hchokshi committed:
Move account order link change into module

hchokshi committed:
Move account order response logic into module

hchokshi committed:
Move Member-Order relation into module

hchokshi committed:
Merge pull request #22 from hchokshi/feature/member-order-relation-to-module

Member-order relation to module

hchokshi committed:
Add default title to DBAddress scaffolded field

hchokshi committed:
Add tests for DBAddress

hchokshi committed:
Merge pull request #23 from hchokshi/fix/dbaddress-title

Fix DBAddress title, add tests

tractorcow commented on README.md:
```diff
@@ -1,4 +1,6 @@
 # Swipestripe Ecommerce Module
+[![Build Status](https://travis-ci.org/littlegiant/silverstripe-swipestripe.svg?branch=5.0)](https://travis-ci.org/littlegiant/silverstripe-swipestripe)
```
Use the upstream vendor module path here. It'll just get left in when merged up otherwise.

tractorcow commented on _config/extensions.yml:
```diff
@@ -0,0 +1,9 @@
+---
+Name: extensions
+After:
+- framework/*
```
After is unnecessary: Framework doesn't add any extensions to Payment model.

tractorcow commented on _config/injector.yml:
```diff
@@ -0,0 +1,16 @@
+---
+Name: injector
```
Far too generic a name. All names must be unique per project. Use `swipestripe-injector` or something more reliably unique.

tractorcow commented on _config/extensions.yml:
```diff
@@ -0,0 +1,9 @@
+---
+Name: extensions
```
Too generic a name. Use `swipestripe-extensions` instead.

tractorcow commented on composer.json:
```diff
@@ -1,41 +1,76 @@
+        "silverstripe/versioned": "^1.2",
+        "silverstripe/cms": "^4.2",
+        "silverstripe/silverstripe-omnipay": "^3",
+        "heyday/silverstripe-wkhtml": "dev-feature/upgrade-to-ss4"
```
I would suggest to put this into a `swipestripe-pdfrecipts` module to reduce dependencies on non-silverstripe vendors. There's less chance for vendors to drop support and break our core module. :)

tractorcow commented on composer.json:
```diff
@@ -1,41 +1,76 @@
+    "repositories": [
+        {
+            "type": "vcs",
+            "url": "https://github.com/hchokshi/silverstripe-wkhtml.git"
```
Instead of coding this into composer.json, use a custom command in `.travis.yml` to setup.

```bash
composer config repositories.wkhtml vcs https://github.com/hchokshi/silverstripe-wkhtml.git
```

tractorcow commented on composer.lock:
```diff
@@ -0,0 +1,5018 @@
+{
```
Don't add lock file to source control for modules.

tractorcow commented on src/CMSHelper.php:
```diff
@@ -0,0 +1,59 @@
+ * Class CMSHelper
+ * @package SwipeStripe
+ */
+class CMSHelper
```
This probably doesn't need to be a fully injectable service. I would recommend switching this into a trait instead for simplicity / speed. ;)

hchokshi committed:
Make read-only GridField injectable service

tractorcow commented on src/CMSHelper.php:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        foreach ($fieldNames as $fieldName) {
+            $replacement = ReadOnlyGridField::create($fieldList->dataFieldByName($fieldName));
```
We should use performReadonlyTransformation() on 4.2.
4.2.2 will come out soon which will include this fix. We don't need to wait until then; Just use the method raw for the time being.

If we need this in a project we can just use 4.2.x-dev until the tag is released.

You don't need to copy this into the project or try to re-implement it.

tractorcow replied:
Reference https://github.com/silverstripe/silverstripe-framework/pull/8348

hchokshi committed:
Propagate read-only GridField to detail form

hchokshi committed:
Use read-only has one button to view full purchasable details

tractorcow commented on src/Forms/Fields/CountryDropdownField.php:
```diff
@@ -0,0 +1,26 @@
+    public function __construct(string $name, ?string $title = null, $source = [], $value = null)
+    {
+        $source = empty($source)
+            ? IntlLocales::singleton()->getCountries()
```
`i18n::getData()->getCountries()` will respect injected localisation source.

hchokshi committed:
Convert CMSHelper to a trait

tractorcow commented on src/Forms/Fields/ReadOnlyGridField.php:
```diff
@@ -0,0 +1,44 @@
+ * Class ReadOnlyGridField
+ * @package SwipeStripe\Forms\Fields
+ */
+class ReadOnlyGridField extends GridField
```
As discussed, no separate class, please use the core performReadonlyTransformation() method which will use a correct readonly component in 4.2.2.

tractorcow commented on src/GlobalTemplateHelper.php:
```diff
@@ -0,0 +1,82 @@
+    public function ActiveCart(): Order
+    {
+        /** @var Controller|HasActiveCart $controller */
+        $controller = Controller::curr();
```
You can't be sure the current controller has the `HasActiveCart` trait. Instead, add the trait to this class, and implement getRequest() (you can copy Director::currentRequest() for the logic).

tractorcow commented on src/HasActiveCart.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function getActiveCart(): Order
+    {
+        $session = $this->getRequest()->getSession();
```
Can add getRequest() as abstract method instead of `@mixin Controller`

tractorcow commented on src/HasActiveCart.php:
```diff
@@ -0,0 +1,48 @@
+        $cartObj->IsCart = true;
+        $cartObj->write();
+
+        $session->set(SessionData::CART_ID, $cartObj->ID)->save($this->getRequest());
```
Don't save the session. This is handled by session middleware, and should not be invoked directly by user code.

tractorcow commented on src/ORM/FieldType/DBAddress.php:
```diff
@@ -0,0 +1,132 @@
+        'City'     => DBVarchar::class,
+        'Region'   => DBVarchar::class,
+        'Postcode' => DBVarchar::class,
+        'Country'  => DBVarchar::class,
```
You don't need varchar(255) for country code. It should be 3.

Use string notation not object::class notation for dbfields. Use ::class notation only for dataobject relations.

tractorcow commented on src/ORM/FieldType/DBAddress.php:
```diff
@@ -0,0 +1,132 @@
+        'Suburb'   => DBVarchar::class,
+        'City'     => DBVarchar::class,
+        'Region'   => DBVarchar::class,
+        'Postcode' => DBVarchar::class,
```
Can use short postcode; Don't need 255 characters.

tractorcow commented on src/ORM/FieldType/DBAddress.php:
```diff
@@ -0,0 +1,132 @@
+     * @param DBAddress $other
+     * @return $this
+     */
+    public function copyFrom(DBAddress $other): self
```
This is just a re-implementation of setValue(). setValue() accepts another dbcomposite object.

tractorcow commented on src/ORM/FieldType/DBAddress.php:
```diff
@@ -0,0 +1,132 @@
+     * @param string $fieldName
+     * @return DBAddress
+     */
+    public function copyFromArray(array $data, string $fieldName = ''): self
```
Same as above: setValue() also supports array.

tractorcow commented on src/ORM/FieldType/DBAddress.php:
```diff
@@ -0,0 +1,132 @@
+     */
+    public function Empty(): bool
+    {
+        return empty($this->Unit) && empty($this->Street) && empty($this->Suburb) && empty($this->City) &&
```
Can't you just loop over `$this->compositeDatabaseFields()` instead of listing them individually?

tractorcow commented on src/ORM/FieldType/DBAddress.php:
```diff
@@ -0,0 +1,132 @@
+    public function scaffoldFormField($title = null, $params = null)
+    {
+        return FieldGroup::create($title ?? FieldGroup::name_to_label($this->getName()), [
+            TextField::create("{$this->getName()}Unit", 'Unit'),
```
Loop over fields instead of listing one at a time. You can add an exception in the loop for the country field.
You should use scaffoldFormField() on the individual nested DBFields instead of assuming TextField.

tractorcow replied:
I suggest using localisation to get the form field names instead of hard-coding the values.

E.g. `_t(self::class . '_' . $field, $field);`

And put the localised values / defaults in `lang/en.yml`

tractorcow commented on src/Order/AddOnPriority.php:
```diff
@@ -0,0 +1,15 @@
+ * Interface AddOnPriority
+ * @package SwipeStripe\Order
+ */
+interface AddOnPriority
```
Not really needed. We can just use integers.

tractorcow commented on src/Order/Cart/CartForm.php:
```diff
@@ -0,0 +1,114 @@
+ */
+class CartForm extends Form
+{
+    const QUANTITY_FIELD_PATTERN = 'Qty_%d';
```
If these constants are only ever used in this file (as far as I can tell) you probably don't need them.

tractorcow commented on src/Order/Cart/CartForm.php:
```diff
@@ -0,0 +1,114 @@
+     */
+    public function UpdateCart(): HTTPResponse
+    {
+        $this->saveInto($this->cart);
```
We should use getCart() instead of direct property access if we can help it.

tractorcow commented on src/Order/Cart/CartForm.php:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * @return HTTPResponse
+     */
+    public function UpdateCart(): HTTPResponse
```
Action handling should probably be on a FormRequestHandler subclass. Should we implement a `CartFormRequestHandler`, and override `buildRequestHandler()` to build this by default instead of the base class?

tractorcow commented on src/Order/Cart/CartForm.php:
```diff
@@ -0,0 +1,114 @@
+     * @param array $data
+     * @return HTTPResponse
+     */
+    public function RemoveOrderItem(array $data): HTTPResponse
```
As above, let's do request handling in a request handler class.

tractorcow commented on src/Order/Cart/CartFormValidator.php:
```diff
@@ -0,0 +1,40 @@
+                'Your cart is currently locked because there is a checkout in progress. Please complete or cancel the checkout process to modify your cart.'));
+        }
+
+        return $this->result->isValid();
```
You should add extension points for other modules (e.g. payment) to add their own validation rules.

tractorcow commented on src/Order/Cart/CartFormValidator.php:
```diff
@@ -0,0 +1,40 @@
+    {
+        if (!$this->form->getCart()->IsMutable()) {
+            $this->result->addError(_t(self::class . '.CART_LOCKED',
+                'Your cart is currently locked because there is a checkout in progress. Please complete or cancel the checkout process to modify your cart.'));
```
Break long text to fit psr-2 max line length guidelines.

tractorcow replied:
Shall we just add this to travis?

tractorcow replied:
Framwork does this in `.travis.yml`


```yml
- if [[ $PHPCS_TEST ]]; then composer run-script lint; fi
```

Then composer has this

```json
"scripts": {
        "lint": "phpcs src/ tests/php/ tests/behat/src/",
        "lint-clean": "phpcbf src/ tests/php/ tests/behat/src/"
    },
```

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+    /**
+     * @var ServiceFactory
+     */
+    public $paymentServiceFactory;
```
Please use `set<Service>(<ServiceType> $service)` / `get<Service>()` instead of public properties.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+     */
+    public function __construct(Order $cart, ?RequestHandler $controller = null, ?string $name = null)
+    {
+        $this->cart = $cart;
```
Should we add a setter.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+                : $defaultMessage;
+        }
+
+        $this->paymentError = $errorMessage;
```
I wonder if a few extension points around this class would make it easier for users to extend this logic?

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+     * @param array $data
+     * @return HTTPResponse
+     * @throws \SilverStripe\Omnipay\Exception\InvalidConfigurationException
+     * @throws \SilverStripe\Omnipay\Exception\InvalidStateException
```
I prefer importing these long namespaces; It makes the code more readable.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+     */
+    public function ConfirmCheckout(array $data): HTTPResponse
+    {
+        $this->cart->Lock();
```
We should be more consistent with using getters where they exist rather than direct property access.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+    {
+        /** @var OrderConfirmationPage $orderConfirmationPage */
+        $orderConfirmationPage = OrderConfirmationPage::get_one(OrderConfirmationPage::class);
+        return $orderConfirmationPage->LinkForOrder($this->cart);
```
Maybe we should have an empty check to prevent calling methods on null.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+            ->setSuccessUrl($this->getSuccessUrl())
+            ->setFailureUrl($this->getFailureUrl($payment));
+
+        if ($payment->isChanged(null, Payment::CHANGE_VALUE)) {
```
Can't you just call $payment->write() and let auto-change detection conditionally write it?

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+        if (!$payment->Identifier) {
+            // Force identifier to be generated
+            /** @see Payment::onBeforeWrite() */
+            $payment->write();
```
Shouldn't you do this in Payment::init() so you don't need to check it later?

hchokshi replied:
This is an issue/limitation of the `silverstripe/silverstripe-omnipay` module.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+
+        $paymentMethod = $data[static::PAYMENT_METHOD_FIELD];
+        $dueMoney = $this->cart->UnpaidTotal()->getMoney();
+        $payment->init($paymentMethod, $this->supportedCurrencies->formatDecimal($dueMoney),
```
I'm not sure we need a single method if the sole purpose is to wrap a trio of setters. Why not just call setGateway() etc directly? You could init the Identifier in setGateway() since it relies on it to generate it.

hchokshi replied:
This is an issue/limitation of the `silverstripe/silverstripe-omnipay` module.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+        $gatewayField = count($gateways) > 1
+            ? OptionsetField::create(static::PAYMENT_METHOD_FIELD,
+                _t(self::class . '.PAYMENT_METHOD', 'Select your payment method'), $gateways)
+            : HiddenField::create(static::PAYMENT_METHOD_FIELD, null, key($gateways));
```
If you have only a single allowed gateway, using a hidden field requires you to validate it. It's better to not have a field at all, since the value becomes a security attack vector.

In your form submission logic you simply check if there are multiple supported options, and only check the submitted value if there are multiple.

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+            $fieldFactory = GatewayFieldsFactory::create($gateway);
+            $fields->merge($fieldFactory->getFields());
+        }
+
```
Suggest extension point?

tractorcow commented on src/Order/Checkout/CheckoutForm.php:
```diff
@@ -0,0 +1,269 @@
+    protected function buildActions(): FieldList
+    {
+        return FieldList::create(
+            FormAction::create('ConfirmCheckout', _t(self::class . '.CONFIRM_CHECKOUT', 'Confirm Checkout'))
```
Suggest extension point?

tractorcow commented on src/Order/Checkout/CheckoutFormRequestHandler.php:
```diff
@@ -0,0 +1,25 @@
+        $response = parent::redirectBack();
+
+        // Strip query string (e.g. previous payment failure)
+        $response->addHeader('Location', strtok($response->getHeader('Location'), '?'));
```
Use ->redirect() instead of manually building the header.

tractorcow commented on src/Order/Checkout/CheckoutFormValidator.php:
```diff
@@ -0,0 +1,94 @@
+        if ($cart->Empty()) {
+            $this->result->addError(_t(self::class . '.CART_EMPTY',
+                'It looks like your cart is empty. Please add some items before attempting to checkout.'));
+        }
```
Not 100% sure but we may need to check if it's locked here as well.

tractorcow commented on src/Order/Order.php:
```diff
@@ -0,0 +1,547 @@
+        'CartLocked'       => DBBoolean::class,
+        'GuestToken'       => DBVarchar::class,
+        'ConfirmationTime' => DBDatetime::class,
+        'Environment'      => DBEnum::class . '(array("' . CoreKernel::DEV . '", "' . CoreKernel::TEST . '", "' . CoreKernel::LIVE . '"), "' . CoreKernel::DEV . '")',
```
Very unsure about this. This has a strong code smell. A site + db should only ever be on one environment.

hchokshi committed:
Merge pull request #25 from hchokshi/feature/admin-view-order-item-details

Feature - admin view full order item details

tractorcow commented on src/Order/Order.php:
```diff
@@ -0,0 +1,547 @@
+     * @var array
+     */
+    private static $extensions = [
+        Payable::class,
```
here's a tip; Add an array key (eg. `'payable'`) and users can override that key in yml. :)

tractorcow commented on src/Order/Order.php:
```diff
@@ -0,0 +1,547 @@
+    /**
+     * @var string
+     */
+    private static $default_sort = 'ConfirmationTime DESC, LastEdited DESC';
```
Suggest to use sql double quotes on sortable fields. Otherwise some DB's complain (e.g. postgres).

```
private static $default_sort = '"ConfirmationTime" DESC, "LastEdited" DESC';
```

tractorcow commented on src/Order/Order.php:
```diff
@@ -0,0 +1,547 @@
+    /**
+     * @var SupportedCurrenciesInterface
+     */
+    public $supportedCurrencies;
```
protected var, public getter / setter with typed arguments please.

tractorcow commented on src/Order/Order.php:
```diff
@@ -0,0 +1,547 @@
+     * @param null|Payment $payment
+     * @throws \Exception
+     */
+    public function paymentCancelled(?Payment $payment): void
```
Could use some phpdoc on some of these methods. Not sure what a lot of these are meant to do.

tractorcow reviewed (commented):
As commented.

hchokshi committed:
Update CI badges to upstream repo

hchokshi committed:
Update YAML config names to be more unique

hchokshi committed:
Move WKHTML fork override to Travis CI config

hchokshi committed:
Use localised country names in CountryDropdownField

hchokshi committed:
Simplify GlobalTemplateHelper current request logic

hchokshi committed:
Remove copyFrom methods in DBAddress

hchokshi committed:
Simplify DBAddress empty logic

hchokshi committed:
Simplify DBAddress field scaffolding logic

hchokshi committed:
Use getters more consistently for extensibility

hchokshi committed:
Use read-only property doc over public properties for DI

hchokshi committed:
Simplify checkout form redirect logic

hchokshi committed:
Remove environment tracking from Order

hchokshi committed:
Use SilverStripe’s read-only GridField functionality

hchokshi committed:
Move checkout form action to request handler

hchokshi committed:
Move cart form actions to request handler

hchokshi committed:
Remove hidden field for singular payment method

hchokshi committed:
Change DBAddress country code column to 2 chars

hchokshi committed:
Use string specs for database fields

hchokshi committed:
Add checks for in-progress checkouts to checkout form

hchokshi committed:
Add form as parameter to checkout form handler extensions

hchokshi committed:
Move Omnipay payment data logic to Order class

hchokshi committed:
Move cart creation logic into Order class

hchokshi committed:
Allow users to checkout locked order by cloning

hchokshi committed:
Fix travis build Github credentials error

hchokshi committed:
Ensure gateway form fields sent to Omnipay

hchokshi committed:
Merge pull request #26 from hchokshi/address-review-comments

Address review comments

hchokshi committed:
Use Permission::checkMember over Permission::check

Avoids having to put ‘any’ arg everywhere - potential pain if 
SilverStripe decide to change the default value in the future.

hchokshi committed:
Add order status field to Order

hchokshi committed:
Add localised title method to Order

hchokshi committed:
Add order status management permission

hchokshi committed:
Add display logic dependency

hchokshi committed:
Allow GridFieldViewButton to be restricted to read-only fields

hchokshi committed:
Add AlwaysModifiableGridField

hchokshi committed:
Add initial implementation of order status updates

hchokshi committed:
Use order title for receipt PDF name

hchokshi committed:
Restrict order update notifications from extensions

hchokshi committed:
Use order title in confirmation email subject

hchokshi committed:
Implement notification emails for order status updates

hchokshi committed:
Show order updates on view order page

hchokshi committed:
Add tests for order status updates

hchokshi committed:
Add mutability and existence checks to setActiveCart

hchokshi committed:
Merge pull request #27 from hchokshi/feature/order-status

Feature - order status updates

hchokshi committed:
Use Order::createCart instead of manual creation

hchokshi committed:
Use setComponent over manually setting fields

hchokshi committed:
Add tests for OrderItemQuantityField

hchokshi committed:
Merge pull request #28 from hchokshi/feature/test-order-qty-field

Add tests for order item quantity field

hchokshi committed:
Allow locked carts to be set as active

hchokshi committed:
Merge pull request #29 from hchokshi/fix/set-immutable-active-cart

Allow locked carts to be set as active

hchokshi committed:
Change YAML prefix to swipestripe-core

hchokshi committed:
Use Lock method rather than setting CartLocked in tests

hchokshi committed:
Add test for OrderItem version locking

hchokshi committed:
Use locking time and versioned to handle version locking

This is more robust than locking to a specific integer version.

hchokshi committed:
Add validation to prevent unlocked non-cart Order

hchokshi committed:
Simplify order lock/unlock behaviour

hchokshi committed:
Update silverstripe/versioned constraint to 1.2.x-dev

1.2.x-dev has fix for relations between staged/unstaged versioned 
objects.

hchokshi committed:
Merge pull request #30 from hchokshi/feature/simplify-version-locking

Simplify version locking behaviour

hchokshi commented:
Review comments have been addressed (till the next review).
